wagtail/models/i18n.py

```diff
@@ -174,13 +174,143 @@ def is_active(self) -> bool:
+        same length as the original queryset.
+
+        By default, the same ordering definition as in the original queryset is applied.
+        This means that the translated values are being considered during ordering,
+        which can lead to a different order than the original queryset. If you want to
+        preserve the same order as the original queryset, you need to pass
+        ``keep_order=True``.
+
+        If a model inherits from :py:class:`DraftStateMixin <wagtail.models.DraftStateMixin>`,
+        draft translations are not considered as translated instances. If a translation
+        is in draft, the original instance is used instead. To override this behavior
+        and include draft translations, pass ``include_draft_translations=True``.
+
+        Note: If localization is disabled via the ``WAGTAIL_I18N_ENABLED`` setting, this
+        method returns the original queryset unchanged.
+
+        """
+        # Skip localization if i18n is not enabled. This behavior is consistent with
+        # the behavior of the `localized` property on `TranslatableMixin`.
+        if not getattr(settings, "WAGTAIL_I18N_ENABLED", False):
+            return self
+
+        # Get all instances that are available in the active locale. We can find these
+        # by getting all model instances that have a translation key from the original
+        # queryset and that are available in the active locale.
+        active_locale = Locale.get_active()
+        original_translation_keys = self.values_list("translation_key", flat=True)
+        translated_instances = self.model.objects.filter(
+            locale_id=pk(active_locale),
+            translation_key__in=original_translation_keys,
+        )
+
+        # Don't consider draft translations. If a translation is in draft, we want to
+        # use the original instance instead. To do so, we exclude draft translations.
+        # This only applies if the model has a `live` field. We allow bypassing this
+        # behavior by passing `include_draft_translations=True`.
+        from wagtail.models import DraftStateMixin
+
+        if issubclass(self.model, DraftStateMixin) and not include_draft_translations:
+            translated_instances = translated_instances.exclude(live=False)
+
+        # Get all instances that are not available in the active locale. We can find
+        # these by excluding the translation keys for which translations exist from the
+        # original queryset.
+        translated_translation_keys = translated_instances.values_list(
+            "translation_key", flat=True
+        )
+        untranslated_instances = self.exclude(
+            translation_key__in=translated_translation_keys,
+        )
+
+        # Combine the two querysets to get the localized queryset.
+        localized_queryset = self.model.objects.filter(
+            models.Q(pk__in=translated_instances)
+            | models.Q(pk__in=untranslated_instances)
+        )
```

Comment:
```suggestion
        translated_vals = translated_instances.values_list(
            "translation_key", "pk"
        )
        translated_translation_keys = (key for key, pk in translated_vals)
        translated_pks = (pk for key, pk in translated_vals)
        untranslated_pks = self.exclude(
            translation_key__in=translated_translation_keys,
        ).values_list("pk", flat=True)

        # Combine the two querysets to get the localized queryset.
        localized_queryset = self.model.objects.filter(
            models.Q(pk__in=translated_pks)
            | models.Q(pk__in=untranslated_pks)
        )
```

Reply:
`translated_vals` in the above suggestion fetches the keys and pks in a single query, meaning `localized_queryset` doesn't need to make another

I've also updated the `untranslated_pks` query to specifically fetch pks, as the entire queryset would have been used before `localised_queryset`

Reply:
I am not sure I follow. The entire localized queryset is already fetched in a single query. The keys and the separate querysets are never actually fetched. They are only used to build the final query. 

I actually think that iterating over the values list would trigger an additional query no? 
```python
        translated_translation_keys = (key for key, pk in translated_vals)
        translated_pks = (pk for key, pk in translated_vals)
```

Reply:
Wow. Django's lazy queryset evaluation truly is on another level these days! It'll be a complicated query for sure, but a single database round-trip, which makes all the difference. Excellelent work @tbrlpld!

Reply:
Thanks @ababic. 

Yea, it's pretty amazing what the resolver can do. I looked at the query once and can definitely confirm it's on the bigger side 😅